src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+################################################################################
+
+"""Fetch models over HTTP."""
+
+import contextlib
+import errno
+import io
+import hashlib
+import re
+import logging
+import os
+import urllib.parse
+from typing import Dict, Tuple, Optional, Type, TypeVar, Any, cast
+from typing_extensions import Protocol
+
+import requests
+import requests_file
+
+from . import models
+
+
+MAX_ALIASES = 30
+_logger = logging.getLogger(__name__)
+_F = TypeVar('_F', bound='Fetcher')
+_M = TypeVar('_M', bound='models.Model')
+
+
+class Session(Protocol):
+    """Generalization of :class:`requests.Session`."""
+
+    def get(self, url: str, **kwargs) -> requests.Response: ...   # pragma: nocover
+    def head(self, url: str, **kwargs) -> requests.Response: ...  # pragma: nocover
+    def close(self) -> None: ...                                  # pragma: nocover
+
+
+class HttpFile(io.RawIOBase):
+    """File-like object that fetches byte ranges via HTTP.
+
+    This requires the server to advertise for support byte-range requests and
+    to provide a Content-Length. It is currently *not* robust against the
+    content changing.
+
+    Raises
+    ------
+    FileNotFoundError
+        HTTP 404 error
+    PermissionError
+        HTTP 403 error
+    OSError
+        Other HTTP errors, or the server doesn't implement the required features.
+    """
+
+    def __init__(self, session: Session, url: str) -> None:
+        self._session = session
+        self._offset = 0
+        # TODO do we need to set Accept-Encoding: none? Not sure how transfer
+        # encoding interact with byte ranges.
+        with session.head(url) as resp:
+            if resp.status_code == 404:
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), url)
+            elif resp.status_code in {401, 403}:
+                raise PermissionError(errno.EACCES, os.strerror(errno.EACCES), url)
+            resp.raise_for_status()
+            if resp.headers.get('Accept-Ranges', 'none') != 'bytes':
+                raise OSError(None, 'Server does not accept byte ranges', url)
+            try:
+                self._length = int(resp.headers['Content-Length'])
+            except (KeyError, ValueError):
+                raise OSError(None, 'Server did not provide Content-Length header', url) from None
+            # TODO: consider storing ETag/Last-Modified to check for data
+            # changing under us.
+            self._url = resp.url
+
+    @property
+    def url(self) -> str:
+        """The actual URL with the content.
+
+        This may differ from the constructor argument if HTTP redirects occurred.
+        """
+        return self._url
+
+    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
+        if whence == io.SEEK_SET:
+            new_offset = offset
+        elif whence == io.SEEK_CUR:
+            new_offset = self._offset + offset
+        elif whence == io.SEEK_END:
```

Comment:
Not sure why SEEK_END is adding the offset, shouldn't this be clamped to the end of the file. Even that doesn't seem useful ? Or are we expecting negative offsets in this case. If so then maybe an out of bounds check to make sure.

Reply:
Adding the offset is how the system call works - you generally use a negative offset (or more commonly zero, to go to the end of the file). From what I can tell it's legal to seek past the end of the file - you just get no data if you try to read there (for a R/W file you can write there, which extends the file).

Reply:
OK, although I still think we might see some strangeness allowing reads rather than refusing to read past the end of the file.